File: langs/csharp/src/VMx/Composites/FilteredCompositeVM.cs
using System.Collections.Specialized;
using VMx.Components;

namespace VMx.Composites;

/// <summary>Cursor behavior when the current item is filtered out.</summary>
public enum FilteredCursorPolicy
{
    /// <summary>Move current to the first visible item, or null when none are visible.</summary>
    SnapToFirst,

    /// <summary>Clear current to null.</summary>
    Clear,

    /// <summary>Keep current only while it remains visible; otherwise clear it.</summary>
    PreserveIfVisible,
}

/// <summary>Maintains a filtered visible projection over a source composite.</summary>
public class FilteredCompositeVM<VM> : IDisposable
    where VM : class, IComponentVM
{
    private readonly Func<VM, bool> _defaultPredicate;
    private readonly FilteredCursorPolicy _cursorPolicy;
    private readonly List<VM> _visible = [];
    private Func<VM, bool> _predicate;
    private VM? _current;
    private bool _disposed;

    /// <summary>Create a filtered projection.</summary>
    public FilteredCompositeVM(
        CompositeVMBase<VM> source,
        Func<VM, bool>? predicate = null,
        FilteredCursorPolicy cursorPolicy = FilteredCursorPolicy.SnapToFirst)
        : this(source, predicate, cursorPolicy, deferInitialRecompute: false)
    {
    }

    /// <summary>Constructor variant used by subclasses that need to initialize first.</summary>
    protected FilteredCompositeVM(
        CompositeVMBase<VM> source,
        Func<VM, bool>? predicate,
        FilteredCursorPolicy cursorPolicy,
        bool deferInitialRecompute)
    {
        Source = source ?? throw new ArgumentNullException(nameof(source));
        _defaultPredicate = _ => true;
        _predicate = predicate ?? _defaultPredicate;
        _cursorPolicy = cursorPolicy;
        Source.CollectionChanged += OnSourceCollectionChanged;
        if (!deferInitialRecompute) Recompute();
    }

    /// <summary>The source composite.</summary>
    protected CompositeVMBase<VM> Source { get; }

    /// <summary>The visible projection snapshot.</summary>
    public IReadOnlyList<VM> Visible => _visible.ToArray();

    /// <summary>Number of visible items.</summary>
    public int VisibleCount => _visible.Count;

    /// <summary>Current item in the visible domain.</summary>
    public VM? Current
    {
        get => _current;
        set => SetCurrent(value);
    }

    /// <summary>Fires whenever projection or current state changes.</summary>
    public event EventHandler? Changed;

    /// <summary>Replace the predicate and recompute the projection.</summary>
    public void SetPredicate(Func<VM, bool> predicate)
    {
        _predicate = predicate ?? throw new ArgumentNullException(nameof(predicate));
        Recompute();
    }

    /// <summary>Set current to null or a currently visible item.</summary>
    public void SetCurrent(VM? item)
    {
        if (item is not null && VisibleIndexOf(item) < 0)
            throw new InvalidOperationException("Current must be null or a visible item.");
        if (ReferenceEquals(_current, item)) return;
        _current = item;
        Changed?.Invoke(this, EventArgs.Empty);
    }

    /// <summary>Move to the next visible item, clamped at the last item.</summary>
    public void MoveToNextVisible()
    {
        if (_visible.Count == 0)
        {
            SetCurrent(null);
            return;
        }
        if (_current is null)
        {
            SetCurrent(_visible[0]);
            return;
        }
        var index = VisibleIndexOf(_current);
        if (index < 0)
        {
            SetCurrent(_visible[0]);
            return;
        }
        SetCurrent(_visible[Math.Min(index + 1, _visible.Count - 1)]);
    }

    /// <summary>Move to the previous visible item, clamped at the first item.</summary>
    public void MoveToPreviousVisible()
    {
        if (_visible.Count == 0)
        {
            SetCurrent(null);
            return;
        }
        if (_current is null)
        {
            SetCurrent(_visible[0]);
            return;
        }
        var index = VisibleIndexOf(_current);
        if (index < 0)
        {
            SetCurrent(_visible[0]);
            return;
        }
        SetCurrent(_visible[Math.Max(index - 1, 0)]);
    }

    /// <summary>Computes the ordered visible projection.</summary>
    protected virtual IReadOnlyList<VM> OrderedVisible() =>
        Source.Where(_predicate).ToArray();

    /// <summary>Recompute projection and reconcile current.</summary>
    protected void Recompute()
    {
        _visible.Clear();
        _visible.AddRange(OrderedVisible());
        if (_current is not null && VisibleIndexOf(_current) < 0)
        {
            _current = _cursorPolicy == FilteredCursorPolicy.SnapToFirst
                ? _visible.FirstOrDefault()
                : null;
        }
        else if (_current is null && _cursorPolicy == FilteredCursorPolicy.SnapToFirst)
        {
            _current = _visible.FirstOrDefault();
        }
        Changed?.Invoke(this, EventArgs.Empty);
    }

    private int VisibleIndexOf(VM item) =>
        _visible.FindIndex(candidate => ReferenceEquals(candidate, item));

    private void OnSourceCollectionChanged(object? sender, NotifyCollectionChangedEventArgs e) => Recompute();

    /// <inheritdoc/>
    public void Dispose()
    {
        if (_disposed) return;
        _disposed = true;
        Source.CollectionChanged -= OnSourceCollectionChanged;
    }
}

File: langs/python/src/vmx/composites/filtered_composite_vm.py
"""FilteredCompositeVM — visible projection over a CompositeVM source."""

from __future__ import annotations

from collections.abc import Callable
from enum import Enum
from typing import Generic, TypeVar

import reactivex as rx
from reactivex import operators as ops
from reactivex.subject import Subject

from vmx.components.base import _ComponentVMBase
from vmx.composites.composite_vm import CompositeVM

VM = TypeVar("VM", bound=_ComponentVMBase)


class FilteredCursorPolicy(str, Enum):
    SNAP_TO_FIRST = "snap_to_first"
    CLEAR = "clear"
    PRESERVE_IF_VISIBLE = "preserve_if_visible"


class FilteredCompositeVM(Generic[VM]):
    """Maintains a filtered visible projection over a source composite."""

    def __init__(
        self,
        source: CompositeVM[VM],
        *,
        predicate: Callable[[VM], bool] | None = None,
        cursor_policy: FilteredCursorPolicy = FilteredCursorPolicy.SNAP_TO_FIRST,
    ) -> None:
        self._source = source
        self._predicate = predicate or (lambda _: True)
        self._cursor_policy = cursor_policy
        self._visible: list[VM] = []
        self._current: VM | None = None
        self._changed: Subject[None] = Subject()
        self._disposed = False
        self._subscription = source.on_collection_changed.subscribe(lambda _: self._recompute())
        self._recompute()

    @property
    def visible(self) -> tuple[VM, ...]:
        return tuple(self._visible)

    @property
    def visible_count(self) -> int:
        return len(self._visible)

    @property
    def current(self) -> VM | None:
        return self._current

    @current.setter
    def current(self, value: VM | None) -> None:
        self.set_current(value)

    @property
    def on_changed(self) -> rx.Observable[None]:
        return self._changed.pipe(ops.as_observable())

    def set_predicate(self, predicate: Callable[[VM], bool]) -> None:
        self._predicate = predicate
        self._recompute()

    def set_current(self, item: VM | None) -> None:
        if self._disposed:
            return
        if item is not None and self._visible_index(item) < 0:
            raise ValueError("current must be None or a visible item")
        if self._current is item:
            return
        self._current = item
        self._changed.on_next(None)

    def move_to_next_visible(self) -> None:
        if not self._visible:
            self.set_current(None)
            return
        index = -1 if self._current is None else self._visible_index(self._current)
        if index < 0:
            self.set_current(self._visible[0])
            return
        self.set_current(self._visible[min(index + 1, len(self._visible) - 1)])

    def move_to_previous_visible(self) -> None:
        if not self._visible:
            self.set_current(None)
            return
        index = -1 if self._current is None else self._visible_index(self._current)
        if index < 0:
            self.set_current(self._visible[0])
            return
        self.set_current(self._visible[max(index - 1, 0)])

    def _ordered_visible(self) -> list[VM]:
        return [item for item in self._source if self._predicate(item)]

    def _recompute(self) -> None:
        if self._disposed:
            return
        self._visible = self._ordered_visible()
        if self._current is None or self._visible_index(self._current) < 0:
            if self._cursor_policy in (
                FilteredCursorPolicy.CLEAR,
                FilteredCursorPolicy.PRESERVE_IF_VISIBLE,
            ):
                self._current = None
            else:
                self._current = self._visible[0] if self._visible else None
        self._changed.on_next(None)

    def _visible_index(self, item: VM) -> int:
        return next(
            (index for index, candidate in enumerate(self._visible) if candidate is item),
            -1,
        )

    def dispose(self) -> None:
        if self._disposed:
            return
        self._disposed = True
        self._subscription.dispose()
        self._changed.on_completed()
        self._changed.dispose()
